### PR TITLE
Validate the result of a deserialization of a Layout JSON file.

### PR DESCRIPTION
**Pull request by rogermichoud**
The deserialization used by Gson use reflection to recreate an object. There is no validation there, it means that we can construct Layout with some null fields.

The problem got exposed with the addition of UnresponsiveServers list. If you would omit it on the JSON layout file, it will create a Layout alright but throw a NullPointerException later when the management server was trying to access it.

We need to ensure that we create a valid Layout when we deserialize it.

### CODE REVIEW

**Comment by rogermichoud** [comment]
#546

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **e6a073a**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #546 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/615iUiIVSMYaR72g4SBZ3FhFK0yVv3cI)

File: runtime/src/main/java/org/corfudb/runtime/view/Layout.java [hunk 1]
```diff
@@ -71,12 +75,25 @@
 
     transient ConcurrentHashMap<LayoutSegment, AbstractReplicationView> replicationViewCache;
 
+    /* Defensive constructor since we can create a Layout from a JSON file. JSON deserialize is forced through
+     * this constructor.
+     */
     public Layout(List<String> layoutServers, List<String> sequencers, List<LayoutSegment> segments, List<String> unresponsiveServers, long epoch) {
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/Layout.java]: Could you use lombok's @NonNull statement here?

File: runtime/src/main/java/org/corfudb/runtime/view/Layout.java [hunk 2]
```diff
@@ -461,8 +480,11 @@ public LayoutSegment(ReplicationMode replicationMode, long start, long end, List
 
     @Data
     @Getter
-    @AllArgsConstructor
     public static class LayoutStripe {
         final List<String> logServers;
+
+        public LayoutStripe(List<String> logServers) {
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/Layout.java]: Same with using @Nonnull here.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **76a057f**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #546 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/dBPIL8M176V2sPupgo5s5BSueCffzEJY)

**Comment by corfudb-performance** [comment]
Results automatically generated by CorfuDB Benchmark Framework to assess the performance of this pull request for commit 4461598 .

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode

An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
[Pull Request #546 Graphs](https://snapshot.raintank.io/dashboard/snapshot/y4deIVxfkUhQU0Esgw8CUf2FKJ4SnvxA)

**Pull request merged**